File: packages/engine/src/cpus/6502/instructions/brk.js
import { CycleType } from '..'

/**
 * BReaK
 *
 * Affects Flags: B
 *
 * push PC, push P, JMP (0xFFFF/E)
 *
 * Affects Flags: B
 *
 * @example
 * MODE          SYNTAX        HEX LEN TIM
 * Implied       BRK           $00  1   7
 */
export function* brk() {
  this.breakCommand = true
  this.fetch()
  yield CycleType.GET

  this.write(0x100 | this.sp, this.pc >>> 8)
  if (--this.sp < 0) this.sp = 0xff
  yield CycleType.PUT

  this.write(0x100 | this.sp, this.pc & 0xff)
  if (--this.sp < 0) this.sp = 0xff
  yield CycleType.PUT

  const nmi = this.nmiRequested

  this.write(0x100 | this.sp, this.status | 0x10)
  if (--this.sp < 0) this.sp = 0xff
  yield CycleType.PUT

  if (nmi) {
    this.pc &= ~0xff
    this.pc |= this.read(0xfffa)
    this.interruptDisabled = true
    yield CycleType.GET
    this.pc &= 0xff
    this.pc <<= 8
    this.pc |= this.read(0xfffb)
    yield CycleType.GET
  } else {
    this.pc &= ~0xff
    this.pc |= this.read(0xfffe)
    this.interruptDisabled = true
    yield CycleType.GET
    this.pc &= 0xff
    this.pc <<= 8
    this.pc |= this.read(0xffff)
    yield CycleType.GET
  }
}

File: packages/engine/src/cpus/6502/instructions/brk.zig
const std = @import("std");
const CPU = @import("../cpu.zig").CPU;
const Addressing = @import("../opcode.zig").Addressing;

// * BReaK
// *
// * Affects Flags: B
// *
// * push PC, push P, JMP (0xFFFF/E)
// *
// * Affects Flags: B
// *
// * MODE          SYNTAX        HEX LEN TIM
// * Implied       BRK           $00  1   7

pub fn brk(self: *CPU, addressing: Addressing) void {
    _ = addressing;
    if (self.cycle_counter == 2) {
        self.b = true;
        _ = self.fetch();
        self.next_cycle = .write;
        return;
    } else if (self.cycle_counter == 3) {
        self.write(0x100 | @as(u16, self.sp), @truncate(self.pc >> 8));
        self.sp -%= 1;
        self.next_cycle = .write;
        return;
    } else if (self.cycle_counter == 4) {
        self.write(0x100 | @as(u16, self.sp), @as(u8, @truncate(self.pc & 0xff)));
        self.sp -%= 1;
        self.next_cycle = .write;
        return;
    } else if (self.cycle_counter == 5) {
        self.write(0x100 | @as(u16, self.sp), self.getP() | 0x10);
        self.sp -%= 1;
        self.next_cycle = .read;
        return;
    } else if (self.cycle_counter == 6) {
        self.pc &= @as(u16, 0xff00);
        self.pc |= self.read(if (self.nmi_requested) 0xfffa else 0xfffe);
        self.i = true;
        self.next_cycle = .read;
        return;
    } else {
        self.pc &= 0xff;
        self.pc <<= 8;
        self.pc |= self.read(if (self.nmi_requested) 0xfffb else 0xffff);
    }
}

test {
    _ = @import("brk.test.zig");
}
